src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,165 @@
+
+/-- The inclusion map of derivations at a point into the total bundle. -/
+def derivation_inclusion {x : M} (v : point_derivation I x) : derivation_bundle I M :=
+sigma.mk x v
+
+instance [inhabited M] : inhabited (derivation_bundle I M) :=
+⟨derivation_inclusion (0 : point_derivation I (default M))⟩
+
+section
+
+/- Why do I need to rewrite extensionality rules for reducible defs? -/
+namespace point_derivation
+
+variables {I} {M} {x y : M} {v w : point_derivation I x} (f g : C^∞⟮I, M; 𝕜⟯) (r : 𝕜)
+
+lemma coe_injective (h : ⇑v = w) : v = w :=
+@derivation.coe_injective 𝕜 _ C^∞⟮I, M; 𝕜⟯ _ _ 𝕜 _ (point_derivation.module I x) _
+  (point_derivation.is_scalar_tower I x) v w h
+
+@[ext] theorem ext (h : ∀ f, v f = w f) : v = w :=
+coe_injective $ funext h
+
+theorem hext {u : point_derivation I y} (h1 : x = y) (h2 : ∀ f, v f = u f) : v == u :=
+by { cases h1, rw heq_iff_eq, ext, exact h2 f }
+
+end point_derivation
+
+end
+
+section
+
+variables (I) {M} (X Y : derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯)
+  (f g : C^∞⟮I, M; 𝕜⟯) (r : 𝕜)
+
+/-- Evaluation at a point gives rise to a `C^∞⟮I, M; 𝕜⟯`-linear map between `C^∞⟮I, M; 𝕜⟯` and `𝕜`.
+ -/
+def smooth_function.eval_at (x : M) :
+  @linear_map C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ 𝕜 _ _ _ _ (point_derivation.module I x) :=
+@alg_hom.to_linear_map C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ 𝕜 _ _ _ _ (point_derivation.algebra I x)
+  (point_derivation.smooth_function.eval I x)
+
+namespace derivation
+
+variable {I}
+
+/-- The evaluation at a point as a linear map. -/
+def eval_at (x : M) : (derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯) →ₗ[𝕜] point_derivation I x :=
+@linear_map.comp_der 𝕜 _ C^∞⟮I, M; 𝕜⟯ _ _ C^∞⟮I, M; 𝕜⟯ _ _ _ _ 𝕜 _ (point_derivation.module I x) _
+  (point_derivation.is_scalar_tower I x) (smooth_function.eval_at I x)
+
+lemma eval_apply (x : M) : eval_at x X f = (X f) x := rfl
+
+end derivation
+
+variables {I} {E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{H' : Type*} [topological_space H'] {I' : model_with_corners 𝕜 E' H'}
+{M' : Type*} [topological_space M'] [charted_space H' M']
+
+/-- The differential of a function interpreted in the context of derivations. -/
+def fdifferential_map (f : C^∞⟮I, M; I', M'⟯) (x : M) (v : point_derivation I x) :
+  (point_derivation I' (f x)) :=
+{ to_fun := λ g, v (g.comp f),
+  map_add' := λ g h, by rw [smooth_map.add_comp, derivation.map_add],
+  map_smul' := λ k g, by rw [smooth_map.smul_comp, derivation.map_smul],
+  leibniz' := λ g h, by {simp only [derivation.leibniz, smooth_map.mul_comp], refl} }
+
+/-- The differential is a linear map. -/
+def fdifferential (f : C^∞⟮I, M; I', M'⟯) (x : M) : (point_derivation I x) →ₗ[𝕜]
+  (point_derivation I' (f x)) :=
+{ to_fun := fdifferential_map f x,
+  map_smul' := λ k v, rfl,
+  map_add' := λ v w, rfl }
+
+/- Standard notion for the differential. The abbreviation is `MId`. -/
+localized "notation `𝒅` := fdifferential" in manifold
+
+lemma apply_fdifferential (f : C^∞⟮I, M; I', M'⟯) (x : M) (v : point_derivation I x)
```

Comment:
Should this be `@[simp]`?

Reply:
I deliberately did not make it simp because `g : C^∞⟮I', M'; 𝕜⟯` will always be there in this theory to avoid `heq` problems, and one will often not want to rewrite this lemma, so that other lemmas dealing with the differential can see its presence

Reply:
I'm afraid I don't follow the point about `heq`. Could you expand?

Reply:
In order to avoid comparing two point derivations at two equal points that are not definitionally equal (it happens all the time) equalities will always be written applied to a universally quantified function.

Reply:
I still don't understand the point you're making @Nicknamen

Reply:
`heq` appears nowhere precisely because by embedding extensionality in the equalities all `heq` are avoided. When I wrote `heq` problems I actually meant `heq`. I mean `heq` is itself a problem so it is better to avoid it, or at least I was told so on Zulip.

Reply:
I get the gist but I the detail of why this should not be a `simp` lemma is still not clear to me.

I don't doubt that you are correct but given that Scott and Eric were at least briefly unclear on this point, and I remain unclear, I think it would be good to _give an explicit example_ of the problem, using actual Lean code.

I also wonder if we should add a Library Note explaining since it seems you are following a design pattern which deserves explanation.

Reply:
What problem should I give an example of? An example of where I would not want the differential to be simplified is when it appears as
`(𝒅(𝑳 I g)) (1 : G) (eval_at (1 : G) X) f`
with left invariant derivations, because I want to turn this into
`eval_at g X f`
thanks to the left invariant property and I want to sate equalities in terms of the differential and not in terms of its applied version, so to be closer to standard mathematics.
If I ran simpl here with `apply` as `simp`, I would not be able to rewrite the left invariant property anymore.

Reply:
Thanks, I think this is exactly the sort of example I was hoping for. Do you think you could turn this into a self-contained example? I.e., a file that begins `import geometry.manifold.derivation_bundle`.

For example I just tried to explore your example by simple-mindedly pasting the code beneath `apply_fdifferential` and it took me a while before I realised:

- I need to `open_locale lie_group` to get the `𝑳` notation to work.
- The `g` in your code is an element of the group `G` (whereas it is a smooth function on `M'` in the lemma `apply_fdifferential` which we are discussing).
- I need to define `G` and give it a Lie group structure for this example to work.

I appreciate you must be sick of discussing what seems like a small point but I feel that if you could provide a self-contained example with something that works when you do not have this lemma marked as `simp` and which fails when you do, then we could clear this up quickly.

Reply:
Hey, no worries, I am not sick, but I thought there was no need to include a working example because I quoted left invariant derivations precisely so that one could look at that file for multiple real-life examples. If you look for instances of `left_invariant_derivation` there you will find the self contained examples you are looking for! Does this answer your question or do you still prefer me to include here some specific lines? I believe there you can see them in the context of a file you already probably read, so you might prefer that?

Reply:
You can find that file by clicking the commit `Removing left invariant derivations`